databricks/koalas/frame.py

```diff
@@ -1284,6 +1284,49 @@ def notnull(self):
 
     notna = notnull
 
+    def nunique(self, axis: int = 0, dropna: bool = True) -> 'ks.Series':
+        """
+        Return number of unique elements in the object.
+
+        Excludes NA values by default.
+
+        Parameters
+        ----------
+        axis : Can only be set to 0 at the moment.
+        dropna : Don’t include NaN in the count.
+
+        Returns
+        -------
+        The number of unique values per column as a Koalas Series.
+
+        Examples
+        --------
+        >>> ks.DataFrame({'A': [1, 2, 3], 'B': [np.nan, 3, np.nan]}).nunique()
+        A    3
+        B    1
+        Name: 0, dtype: int64
+
+        >>> ks.DataFrame({'A': [1, 2, 3], 'B': [np.nan, 3, np.nan]}).nunique(dropna=False)
+        A    3
+        B    2
+        Name: 0, dtype: int64
+        """
+        if axis != 0:
+            raise ValueError("The 'nunique' method only works with axis=0 at the moment")
+        if dropna:
+            res = self._sdf.select([F.countDistinct(Column(c))
+                                   .alias(c)
+                                    for c in self.columns])
+        else:
+            res = self._sdf.select([(F.countDistinct(Column(c))
+                                     + F.when(F.count(F.when(F.col(c).isNull(), 1).otherwise(None))
+                                              >= 1, 1).otherwise(0))
+                                   .alias(c)
+                                    for c in self.columns])
+        # TODO: Find a way to avoid converting to pandas and back
+        res = res.toPandas().T.iloc[:, 0]
```

Comment:
hm, out of curiosity, why converting to Pandas is required? I thought we can just return the Spark column without conversion.

Reply:
The problem is that the select statement returns
```
   A  B
0  3  1
```

However, the expected output of the method is
```
A    3
B    1
Name: 0, dtype: int64
```

Do you know a practical way to get from the first to the latter using only PySpark?

Reply:
Yes, we need to perform a transpose here. It should not matter performance-wise though, because there is only one row of data to convert. This will break the lineage in spark but we can always add a pure spark transpose later.

Reply:
Based on the comments below, `DataFrame.nunique()` now returns a pandas Series which solves the conversion issue.

Reply:
Okie, but why don't we convert Pandas Series into Koalas Series as it was done before? It'd be nicer if we match the return type into Koalas series.

Reply:
Do you have good reasons for returning a Koalas Series @HyukjinKwon? Personally, I find @thunterdb's reference to the design principles (see comment below) quite compelling.

Reply:
Since the number of columns is small, it'd make sense to follow our own principle to just return a pandas Series.

Note that an alternative design (which would require us changing the principle and every other place ...) is to always return a koalas Series everywhere.